### ISSUE TITLE
[java] A constant utility class, as described in Effective Java, should not trigger a MissingStaticMethodInNonInstantiatableClass violation

### ISSUE THREAD

**Issue by jbisotti** [pmd/pmd#5197]
**Affects PMD Version:**  7.5.0 (likely earlier 7.x versions)

**Rule:** MissingStaticMethodInNonInstantiatableClass

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/pmd/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass>

**Description:** Our legacy app makes use of the constant utility class pattern, as described by Bloch in Effective Java, 3rd Edition, Item 22. Not all non-instantiable classes require static methods. This is not a problem with 6.55.0 and it is one of the issues preventing us from upgrading to 7.x.

**Code Sample demonstrating the issue:**

```java
public final class FooConstants {

    private FooConstants() {
        throw new UnsupportedOperationException("do not call");
    }

    public static final class Bar {
        public static final String ID = "id";
        public static final String NAME = "name";

        private Bar() {
            throw new UnsupportedOperationException("do not call");
        }
    }

    public static final class Blah {
        public static final String ENTITY_ID = "entityId";
        public static final String EMPLOYEE_STATUS_CODE = "employeeStatusCode";
        public static final String TAX_ID = "taxId";

        private Blah() {
            throw new UnsupportedOperationException("do not call");
        }
    }
}
```

**Expected outcome:** This rule should not object to this pattern.

PMD reports a violation at line 1, but that's wrong. That's a false positive.

**Running PMD through:** *Maven*

**Comment by jsotuyod** [comment]
This is a duplicate of #1652.

The issue is not with the implementation of utility classes (neither `Blah` nor `Bar` from your sample would be flagged by PMD), but with the usage of a class (`FooConstants`) as a namespace, instead of using a Java package for that purpose.

Strictly speaking, the rule is detecting what it's meant to detect. But under #1652 there is some discussion on wether the rule as it stands makes sense or should be replaced with something else. Feel free to join that discussion.
